### PR TITLE
Fix pattern matching bug revealed by newer mypy.

### PR DESCRIPTION
**Pull request by StefanUlbrich**
Unit tests added. 

@alexpovel , thanks to Mypy working, this bug has been revealed. I could confirm the error in a unit test and fix it (making the unit tests passing strict mode once the decorator issue is resolved

### CODE REVIEW

File: .github/workflows/docs_pages.yaml [hunk 1]
```diff
@@ -36,4 +36,4 @@ jobs:
       uses: JamesIves/github-pages-deploy-action@4.1.4
       with:
         branch: gh-pages
-        folder: gh-pages
```
Comment (alexpovel) [inline .github/workflows/docs_pages.yaml:39]: This should probably not be part of the commit.

File: .github/workflows/pull_requests.yaml [hunk 2]
```diff
@@ -5,8 +5,7 @@ on: [push, pull_request]
 jobs:
   test:
     runs-on: ubuntu-latest
-
-  steps:
+    steps:
```
Comment (alexpovel) [inline .github/workflows/pull_requests.yaml:8]: This should probably not be part of the commit either.

File: doc/conf.py [hunk 3]
```diff
@@ -64,4 +64,4 @@
 # Add any paths that contain custom static files (such as style sheets) here,
 # relative to this directory. They are copied after the builtin static files,
 # so a file named "default.css" will overwrite the builtin "default.css".
-html_static_path = ['_static']
\ No newline at end of file
+html_static_path = ['_static']
```
Comment (alexpovel) [inline doc/conf.py:67]: This should probably not be part of the commit either.

File: src/design_by_contract.py [hunk 4]
```diff
@@ -25,14 +25,14 @@ class UnresolvedSymbol:
...
     """
 
-    name: Optional[str]
+    name: str
     value: Optional[Any] = None
 
     def __eq__(self, other: Any) -> Union["UnresolvedSymbol", bool]:  # type: ignore[override]
         match other:
-            case UnresolvedSymbol(None):
+            case UnresolvedSymbol(name, None):
```
Comment (alexpovel) [inline src/design_by_contract.py:33]: Looks fine. An alternative would be named arguments:

```python
case UnresolvedSymbol(name=name, value=None)
```

That would've caught the initial bug instantly without `mypy` since it would've revealed the unintentional `name=None`.

**Review by alexpovel — COMMENTED** [review]
I've never seen/used `mypy` on tests, but why not! Tests are vital and can be buggy. Only issue I see (probably a big one) is that `pytest.parametrize` and similar decorators won't play nice. You're not currently using them so 🤷

File: src/design_by_contract.py (hunk 4 above)
Comment (StefanUlbrich) [reply src/design_by_contract.py:33]: True. Then I would do it for all cases.. will it work with the `_`  (2nd case)?

Comment (StefanUlbrich) [reply src/design_by_contract.py]: > I've never seen/used mypy on tests, but why not! Tests are vital and can be buggy. Only issue I see (probably a big one) is that pytest.parametrize and similar decorators won't play nice. You're not currently using them so shrug

I set up VSCode to also run mypy on every open file. Without these fixes my screen is just red

Comment (StefanUlbrich) [reply src/design_by_contract.py]: > True. Then I would do it for all cases.. will it work with the `_` (2nd case)?

It did .. I adjusted

**Review by alexpovel — APPROVED** [review]
